### PR TITLE
Docs: Polars GroupBy

### LINKED ISSUE #1835: Docs: Grouping + Agg Notebook

### PR DESCRIPTION
**Pull request by GKD-stack**
closes #1835

### CODE REVIEW

File: docs/source/getting-started/tabular-data/agg_and_filter.ipynb [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+    "plt.grid(True)\n",
+    "plt.tight_layout()  \n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "While the differentially computed values still follow the general trend of the actual values, they do vary more significantly due to the smaller sizes of the year-quarter groups. "
```
Comment (mccalluc) [inline docs/source/getting-started/tabular-data/agg_and_filter.ipynb:375]:
```suggestion
    "While the differentially private values still follow the general trend of the actual values, they do vary significantly more due to the smaller sizes of the year-quarter groups. "
```
Re-reading this, it feels like there are really two distinct comparisons that could be made:
- dp vs. actual values, grouped by quarter
- dp values grouped by quarter, vs. dp values grouped by year

There might be something to say here like... while the signal is always going to get weaker with smaller samples, applying DP compounds this, limiting the utility of DP when applied to fine-grained aggregations... or whatever makes sense to you.

File: docs/source/getting-started/tabular-data/agg_and_filter.ipynb [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+      "Actual Total Count of Laid Off Respondents:  76009\n",
+      "DP Total Count of Laid Off Respondents:  76115\n"
+     ]
+    }
+   ],
+   "source": [
+    "filt_actual = (df.filter(pl.col(\"ILOSTAT\")==3).select(pl.len())).collect().item()\n",
+    "print(\"Actual Total Count of Laid Off Respondents: \", filt_actual)\n",
+    "filt_dp = (context.query().filter(pl.col(\"ILOSTAT\")==3).select(pl.len().dp.noise())).release().collect().item()\n",
+    "print(\"DP Total Count of Laid Off Respondents: \", filt_dp)"
```
Comment (mccalluc) [inline docs/source/getting-started/tabular-data/agg_and_filter.ipynb:410]: Feel free to expand the filter section based on the conversation yesterday: Filtering can be seen as a kind of aggregation, and if the result of filtering is a small subset, you'll have a corresponding increase in noise and decrease in utility.

**Review by mccalluc — COMMENTED** [review]
Great structure! Some formatting comments, and a couple places where the discussion could be expanded.

File: docs/source/getting-started/tabular-data/agg_and_filter.ipynb [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this section, you will learn how to compute differentially private statistics while applying key data manipulation techniques such as: \n",
+    "\n",
+    "- Singular Variable Groupby\n",
+    "- Multiple Variable Groupby\n",
+    "- Filtering\n",
+    "\n",
+    "For each method, we will compare the actual values to the computed differentially private values to demonstrate utility. The [documentation](https://docs.opendp.org/en/nightly/api/python/opendp.polars.html#module-opendp.polars) provides more information about the methods. We will use the [sample data](https://github.com/opendp/dp-test-datasets/blob/master/data/sample_FR_LFS.csv.zip) from the Labour Force Survey in France. "
```
[the remark this answers is not in the document]
Comment (GKD-stack) [reply docs/source/getting-started/tabular-data/agg_and_filter.ipynb:20]: chuck you said you'll finish up the last bit here!

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Actual Total Count of Laid Off Respondents:  76009\n",
```
Comment (vikrantsinghal) [inline docs/source/getting-started/tabular-data/agg-and-filter.ipynb:411]: Again, segregate the output from the other stuff via a box or something, whilst indicating explicitly that this is the output.

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+    "    )\n",
+    ").release().collect().item()\n",
+    "print(\"DP Total Count of Laid Off Respondents: \", filt_dp)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Filtering can also be viewed as a type of partitioning. If the result of filtering a smaller subset, it can lead to an increase in noise and decrease in utility. "
```
Comment (vikrantsinghal) [inline docs/source/getting-started/tabular-data/agg-and-filter.ipynb:436]: The last sentence doesn't make much sense.

**Review by vikrantsinghal — DISMISSED** [review]

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+   "metadata": {},
+   "source": [
+    "# Aggregrations and Filtering"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this section, you will learn how to compute differentially private statistics while applying key data manipulation techniques such as: \n",
```
[the remark this answers is not in the document]
Comment (GKD-stack) [reply docs/source/getting-started/tabular-data/agg-and-filter.ipynb:14]: For the compositor?

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
```
[the remark this answers is not in the document]
Comment (GKD-stack) [reply docs/source/getting-started/tabular-data/agg-and-filter.ipynb:54]: Could you explain this a bit further? This line was reading in the csv and there wasn't supposed to be an output, but the loading process does show up

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb (hunk 5 above)
Comment (GKD-stack) [reply docs/source/getting-started/tabular-data/agg-and-filter.ipynb:436]: what about this? 

Similar to partitioning, if the result of filtering is a smaller subset, it can lead to an increase in noise and decrease in utility.

**Comment by mccalluc** [comment]
CI failure:
```
----> 9         ("YEAR", ): dp.Margin(public_info= "keys", max_partition_length=estimated_max_partition_len, max_partition_contributions=4),
     10         ("YEAR", "QUARTER",): dp.Margin(public_info= "keys", max_partition_length=estimated_max_partition_len, max_partition_contributions=1),
     11         (): dp.Margin(public_info= "lengths",max_partition_length=estimated_max_partition_len, max_num_partitions=1),
     12     },
     13 )

AttributeError: module 'opendp.prelude' has no attribute 'Margin'
```
I'll fix this much.

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+    "count_year_actual"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now to get the differentially private statistics, add `dp.noise` after the aggregate function is specified and `.release` after the entire query before `.collect`. \n",
+    "\n",
+    "Calling `.release` is always the final step in compiling your differentially private data in a usable form and ensuring it is compliant with differential privacy guarantees. \n",
```
Comment (chikeabuah) [inline docs/source/getting-started/tabular-data/agg-and-filter.ipynb:176]: I don't think "compiling" is the right word in this context. Also "in a usable form" seems like elusive language. Why is it usable? Usability seems like a different concept from "safety" or "privacy" which is the relevant concept being discussed right here.

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+    "## Multiple Variable *`group_by`* "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To demonstrate the `group_by` method with multiple variables, let's compute the total number of respondents per year and quarter. We'll need a margin for the combination of the variables being used for the grouping. \n",
+    "\n",
+    "In this case, the margin is defined by the `(\"YEAR\", \"QUARTER\",)` entry in the initial definition of `context`. "
```
Comment (chikeabuah) [inline docs/source/getting-started/tabular-data/agg-and-filter.ipynb:285]: I think some prior information on what a margin is, and maybe a link to some documentation would help here.

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb [hunk 10]
```diff
@@ -0,0 +1,481 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "While the differentially private values still follow the general trend of the actual values, they do vary significantly more due to the smaller sizes of the year-quarter groups. \n",
+    "\n",
+    "### Caution Against Excessive Grouping\n",
+    "\n",
+    "Larger data generally corresponds to less noise being added relative to the group size and higher utility. Excessive grouping by multiple keys can lead to smaller partitions. Adding noise to the smaller partitions may dilute the utility of the results, compared to adding noise for larger partitions. \n",
```
Comment (chikeabuah) [inline docs/source/getting-started/tabular-data/agg-and-filter.ipynb:400]: I'm not sure what "larger data" means here, I would reword or remove this first sentence.

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (GKD-stack) [reply docs/source/getting-started/tabular-data/agg-and-filter.ipynb:14]: its all in another introduction, that chuck will link here later

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb (hunk 8 above)
Comment (GKD-stack) [reply docs/source/getting-started/tabular-data/agg-and-filter.ipynb:176]: what do you recommend adding?

**Comment by mccalluc** [comment]
Previewed locally, and made some copy edits. Looking for approval from one more person before merging.

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb [hunk 11]
```diff
@@ -0,0 +1,456 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Aggregrations and Filtering"
```
Comment (Shoeboxam) [inline docs/source/getting-started/tabular-data/agg-and-filter.ipynb:7]:
```suggestion
    "# Group By"
```

Why is filtering in here?

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb [hunk 12]
```diff
@@ -0,0 +1,456 @@
...
+    "plt.grid(True)\n",
+    "plt.tight_layout()  \n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "While the differentially private counts still follow the same trend as the actual values, there is more variance because of the smaller sizes of the year-quarter groups. \n",
```
Comment (Shoeboxam) [inline docs/source/getting-started/tabular-data/agg-and-filter.ipynb:372]: The variance is the same, but for smaller counts the same amount of noise is relatively larger.

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb [hunk 13]
```diff
@@ -0,0 +1,456 @@
...
+    "\n",
+    "### Fine-grained grouping adds more noise\n",
+    "\n",
+    "Larger datasets require less noise to be added to preserve privacy. Grouping by multiple keys can lead to smaller partitions, and adding noise to the smaller partitions may lower the utility of your results. \n",
```
Comment (Shoeboxam) [inline docs/source/getting-started/tabular-data/agg-and-filter.ipynb:376]: This depends on the query. Count queries require a constant amount of noise relative to the dataset size. Grouping by multiple keys can lead to adding the same quantity of noise more times.

File: docs/source/getting-started/tabular-data/agg-and-filter.ipynb [hunk 14]
```diff
@@ -0,0 +1,456 @@
...
+    "\n",
+    "While grouping is an essential data analysis strategy, you'll need to balance the granularity your groups against the accuracy you require."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Filtering"
```
Comment (Shoeboxam) [inline docs/source/getting-started/tabular-data/agg-and-filter.ipynb:385]: This section feels like it belongs in a data pre-processing notebook.

**Review by Shoeboxam — DISMISSED** [review]
Thanks for your work on this. I've left some higher-level feedback that may take some time to incorporate.

**Comment by Shoeboxam** [comment]
* **#1836** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1836?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @GKD-stack and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: docs/requirements_notebooks.txt [hunk 15]
```diff
@@ -6,4 +6,5 @@ seaborn
 scipy
 scikit-learn
 polars==1.1.0
-pyarrow
\ No newline at end of file
+pyarrow
+hvplot
```
Comment (mccalluc) [inline docs/requirements_notebooks.txt:10]: I had prepared a PR that gets rid of a separate requirements file for notebooks:
- #1880

That doesn't need to move forward, but if it did, we'd probably want to be more conservative about adding new dependencies.

File: docs/source/getting-started/tabular-data/index.rst [hunk 16]
```diff
@@ -100,20 +100,18 @@ The specific methods that will be demonstrated are:
   * Quantiles 
 
 * Grouping
+  * Protected Group Keys
```
Comment (mccalluc) [inline docs/source/getting-started/tabular-data/index.rst:103]:
```suggestion

    * Protected Group Keys
```
Sphinx needs a line break between indent levels for correct rendering.

File: docs/source/getting-started/tabular-data/index.rst [hunk 17]
```diff
@@ -100,20 +100,18 @@ The specific methods that will be demonstrated are:
...
+  * using ``with_columns``
+  * using ``filter`` 
 
-* Data Preparation Limitations 
-
-  * Limitations with ``with_columns``
-  * Limitations with ``filter`` 
-
-This section will explain the limitations and properties of common Polars functions that are unique to their usage in OpenDP. 
+This section explains how to build stable dataframe transformations with Polars.
```
Comment (mccalluc) [inline docs/source/getting-started/tabular-data/index.rst:114]: Would it make sense to use RST toctrees here? I could do that, if you don't have all the installs for the doc build.

File: rust/src/accuracy/polars/mod.rs [hunk 18]
```diff
@@ -66,6 +66,7 @@ where
     lazyframe_utility(&lf, alpha)
 }
 
+#[derive(Clone)]
```
Comment (mccalluc) [inline rust/src/accuracy/polars/mod.rs:69]: I'd prefer not to edit the Rust in this PR, if at all possible: if the examples in the docs rely on changes here, then we should get another release out before we point people to the nightly docs. It's adding more steps.

(But if this is a change we really need, don't let me block!)

**Review by mccalluc — COMMENTED** [review]
Haven't actually looked at the notebook yet, but some comments on the surrounding files.

**Review by mccalluc — COMMENTED** [review]
a few more comments.

File: docs/source/getting-started/tabular-data/group-by.ipynb [hunk 19]
```diff
@@ -0,0 +1,928 @@
...
+    "\n",
+    "Reading the table, the smallest suitable noise scale parameter for the sum is 5762. \n",
+    "This may seem large, but remember this noisy sum will be divided by the length, reducing the variance of the final estimate.\n",
+    "You can divide the sum accuracy estimate by the public partition sizes to get disaggregated accuracy estimates for each partition. \n",
+    "\n",
+    "Be mindful that accuracy estimates for sums and means do not take into account bias introduced from clipping.\n",
+    "However, in this example, since the legend for `HWUSUAL` limits work hours between 0 and 98, \n",
+    "the clipping introduced for differential privacy will not further increase bias.\n",
+    "\n",
+    "When the release is made, it is consistent with generally-known estimates of work hours for men and women."
```
Comment (mccalluc) [inline docs/source/getting-started/tabular-data/group-by.ipynb:832]: I don't think we define the codes here? Maybe add a string column? I'd leave off the "generally-known estimates" part, but whatever you think best.

File: docs/source/getting-started/tabular-data/group-by.ipynb [hunk 20]
```diff
@@ -0,0 +1,928 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "While not used in this section, remember that you can request multiple different statistics in one query. \n",
+    "Batching multiple statistics together can not only result in computational speedups, \n",
+    "but also more balanced allocation of privacy budget across queries, \n",
+    "and can help you avoid spending delta multiple times to release the same protected grouping keys.\n",
```
Comment (mccalluc) [inline docs/source/getting-started/tabular-data/group-by.ipynb:895]: Rephrase "spending delta": Not sure it will make sense to readers.

**Review by mccalluc — COMMENTED** [review]
Last comments!

File: rust/src/accuracy/polars/mod.rs (hunk 18 above)
Comment (Shoeboxam) [reply rust/src/accuracy/polars/mod.rs]: Broke it out into a separate PR. It also killed the commit history here, unfortunately.

File: docs/source/getting-started/tabular-data/group-by.ipynb [hunk 21]
```diff
@@ -0,0 +1,928 @@
...
+    "# Group By\n",
+    "\n",
+    "The OpenDP Library allows you to compute statistics over grouped data.\n",
+    "We'll examine three settings with progressively stronger domain descriptors:\n",
+    "\n",
+    "- Protected group keys\n",
+    "- Public group keys\n",
+    "- Public group lengths\n",
+    "\n",
+    "The [documentation](https://docs.opendp.org/en/nightly/api/python/opendp.polars.html#module-opendp.polars) provides more information about the methods. \n",
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply docs/source/getting-started/tabular-data/group-by.ipynb:16]: Neat how it will resolve through notebook conversion.

File: docs/source/getting-started/tabular-data/group-by.ipynb [hunk 22]
```diff
@@ -0,0 +1,928 @@
...
+   "metadata": {},
+   "source": [
+    "## Protected Group Keys\n",
+    "\n",
+    "Grouping keys themselves can be extremely sensitive. \n",
+    "For example, sharing a noisy count query grouped by social security numbers or credit card numbers \n",
+    "would catastrophically violate the privacy of individuals in the data.\n",
+    "Similarly, re-identification of individuals can be accomplished \n",
+    "with very small combinations of grouping columns that may seem benign\n",
+    "(for reference, see how Latanya Sweeney reidentified the Governor of Massachusetts in healthcare data).\n",
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply docs/source/getting-started/tabular-data/group-by.ipynb:173]: Added a link.

File: docs/source/getting-started/tabular-data/group-by.ipynb [hunk 23]
```diff
@@ -0,0 +1,928 @@
...
+    "    },\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "> It is recommended to only ever create one Context that spans all queries you may make on your data.\n",
+    "> We create a second context here to demonstrate how margins will influence the analysis.\n",
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply docs/source/getting-started/tabular-data/group-by.ipynb:473]: We can do another pass when this gets rst'd

File: docs/source/getting-started/tabular-data/group-by.ipynb [hunk 24]
```diff
@@ -0,0 +1,928 @@
...
+   "source": [
+    "context = dp.Context.compositor(\n",
+    "    data=pl.scan_csv(\"sample_FR_LFS.csv\", ignore_errors=True),\n",
+    "    privacy_unit=dp.unit_of(contributions=36),\n",
+    "    privacy_loss=dp.loss_of(epsilon=1.0 / 3),\n",
+    "    split_evenly_over=1,\n",
+    "    margins={\n",
+    "        # grouping keys by \"YEAR\" and \"QUARTER\" are public information\n",
+    "        (\"YEAR\", \"QUARTER\"): dp.polars.Margin(\n",
+    "            public_info=\"keys\",\n",
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply docs/source/getting-started/tabular-data/group-by.ipynb:462]: I changed the query to illustrate how in this different setting, it is reasonable to have public keys.

File: docs/source/getting-started/tabular-data/group-by.ipynb [hunk 25]
```diff
@@ -0,0 +1,928 @@
...
+    "## Public Group Sizes\n",
+    "\n",
+    "It is also possible to declare partition sizes as public information.\n",
+    "Setting this value implies that partition keys are public, \n",
+    "and thus implies an even greater risk of damaging the integrity of the privacy guarantee \n",
+    "than if you were to just specify grouping keys as public.\n",
+    "\n",
+    "Nevertheless, this approach has seen use in high-profile data releases, \n",
+    "including the US Census Bureau in the form of \"data invariants\". \n",
+    "You can think of this as the equivalent of using bounded-DP in each partition.\n",
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply docs/source/getting-started/tabular-data/group-by.ipynb:714]: I just dropped it.

File: docs/source/getting-started/tabular-data/index.rst (hunk 17 above)
Comment (Shoeboxam) [reply docs/source/getting-started/tabular-data/index.rst:115]: Maybe we could do this once the misc notebooks are merged?

File: rust/src/accuracy/polars/mod.rs (hunk 18 above)
Comment (mccalluc) [reply rust/src/accuracy/polars/mod.rs:69]: Huh: Github seems to be confused. The base PR in the stack is merged, and then checked out `main`, and confirmed that this `#[derive(Clone)]` is in there... So it seems like it shouldn't be marked as a change here, as well? Which makes me wonder about how well this UI is representing other changes in this PR.

I'm going to try diffing this branch with `main` locally, and will see what that looks like.

**Review by mccalluc — COMMENTED** [review]
I think I'm up-to-date locally, but I'm not reproducing the error in the docs build.

**Review by mccalluc — APPROVED** [review]
CI passes - A little confused about the git history, and there will be some followup edits in the un-orphan pass, but this is good to merge.
